rig-core/src/providers/eternalai.rs

```diff
@@ -331,6 +337,68 @@ pub fn get_chain_id(key: &str) -> Option<&str> {
     None
 }
 
+pub async fn get_on_chain_system_prompt(
+    rpc_url: &str,
+    contract_addr: &str,
+    agent_id: c_uint,
+) -> Option<String> {
+    abigen!(
+        SystemPromptManagementContract,
+        r#"
+        [{"inputs": [{"internalType": "uint256", "name": "_agentId", "type": "uint256"}], "name": "getAgentSystemPrompt", "outputs": [{"internalType": "bytes[]", "name": "","type": "bytes[]"}], "stateMutability": "view", "type": "function"}]
+        "#
+    );
+    // Connect to an Ethereum node
+    let provider = Provider::<Http>::try_from(rpc_url).expect("Failed to parse url");
+    let client = Arc::new(provider);
+    let contract_address: Address = contract_addr.parse().expect("invalid contract address");
+    let contract = SystemPromptManagementContract::new(contract_address, client);
+    let system_prompts: Vec<Bytes> = contract
+        .get_agent_system_prompt(U256::from(agent_id))
+        .call()
+        .await
+        .expect("invalid agent system prompt");
+
+    let decoded_strings: Vec<String> = system_prompts
+        .iter()
+        .map(|bytes| {
+            String::from_utf8(bytes.to_vec()).unwrap_or_else(|_| "[Invalid UTF-8]".to_string())
+        })
+        .collect();
+
+    if !decoded_strings.is_empty() {
+        let prompt = decoded_strings[0].clone();
+        println!("system prompt : {}", prompt);
+        return fetch_on_chain_system_prompt(&prompt).await;
+    }
+    None
+}
+
+pub async fn fetch_on_chain_system_prompt(content: &str) -> Option<String> {
```

Comment:
Maybe rename to `fetch_from_ipfs`? IMO this would be both more accurate (we're specifically fetching from IPFS, not any "onchain") and more general (we can fetch anything, not just system prompts)